File: java/src/PostorderTravel_145.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class PostorderTravel_145 {

    private class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        LinkedList<TreeNode> stackForNode = new LinkedList<>();
        LinkedList<Boolean> stackForBoolean = new LinkedList<>();
        while (root != null || stackForNode.size() != 0) {
            while (root != null) {
                stackForBoolean.push(Boolean.FALSE);
                stackForNode.push(root);
                root = root.left;
            }
            if (stackForNode.size() != 0) {
                root = stackForNode.pop();
                Boolean temp = stackForBoolean.pop();
                if (temp) {
                    // 第二次弹出
                    res.add(root.val);
                    // 说明现在已经这个节点及其子节点已经访问完毕 需要从stack中拿出其它的访问
                    root = null;
                } else {
                    stackForNode.push(root);
                    stackForBoolean.push(Boolean.TRUE);
                    root = root.right;
                }
            }
        }
        return res;
    }
}

File: go/classic/postOrderTravel_145.go
package classic

import "container/list"

func postorderTraversal(root *TreeNode) []int {
	res := make([]int, 0)

	stack := list.New()
	// 标识这个node是不是第二次访问
	stackForFlag := list.New()
	for root != nil || stack.Len() > 0 {
		for root != nil {
			stack.PushBack(root)
			stackForFlag.PushBack(false)
			root = root.Left
		}

		if stack.Len() > 0 {
			root = stack.Remove(stack.Back()).(*TreeNode)
			flag := stackForFlag.Remove(stackForFlag.Back()).(bool)

			// 说明是第二次访问 这个时候要访问父亲节点
			if flag {
				res = append(res, root.Val)
				root = nil
			} else {
				// 第一次访问
				stack.PushBack(root)
				stackForFlag.PushBack(true)
				root = root.Right
			}
		}
	}
	return res
}

func postorderTraversalWithRecursion(root *TreeNode) []int {
	res := make([]int, 0)
	recursionPostTravel(root, &res)
	return res
}

func recursionPostTravel(root *TreeNode, res *[]int) {
	if root == nil {
		return
	}
	recursionPostTravel(root.Left, res)
	recursionPostTravel(root.Right, res)
	*res = append(*res, root.Val)
}
